src/path.cr

```diff
@@ -0,0 +1,823 @@
+    join parts
+  end
+
+  # Appends the given *parts* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"].join("bar", "baz")             # => Path["foo/bar/baz"]
+  # Path["foo/"].join(Path["/bar/", "/baz"])   # => Path["foo/bar/baz"]
+  # Path["/foo/"].join("/bar/", "/baz/")       # => Path["/foo/bar/baz/"]
+  # ```
+  def join(parts : Array | Tuple) : Path
+    new_name = String.build do |str|
+      str << @name
+      last_ended_with_separator = @type.ends_with_separator?(@name)
+
+      parts.each_with_index do |part, index|
+        case part
+        when Path
+          part = part.@name
+        else
+          part = part.to_s
+          part.check_no_null_byte
+        end
+
+        byte_start = 0
+        byte_count = part.bytesize
+
+        case {@type.starts_with_separator?(part), last_ended_with_separator}
+        when {true, true}
+          byte_start +=1
+          byte_count -=1
+        when {false, false}
+          str << @type.separator unless str.bytesize == 0
+        end
+
+        last_ended_with_separator = @type.ends_with_separator?(part)
+
+        str.write part.unsafe_byte_slice(byte_start, byte_count)
+      end
+    end
+
+    new_instance new_name
+  end
+
+  # Appends the given *part* to this path and returns the joined path.
+  #
+  # ```
+  # Path["foo"] / "bar" / "baz"      # => Path["foo/bar/baz"]
+  # Path["foo/"] / Path["/bar/baz"] # => Path["foo/bar/baz"]
+  # ```
+  def /(part : Path | String) : Path
+    join(part)
+  end
+
+  # Resolves path *name* in this path's parent directory.
+  #
+  # Raises `Path::Error` if `#parent` is `nil`.
+  def sibling(name : Path | String) : Path?
+    if parent = parent()
+      parent.join(name)
+    else
+      raise Error.new("Can't resolve sibling for a path without parent directory")
+    end
+  end
+
+  # Compares this path to *other*.
+  # The comparison is performed strictly lexically: `foo` and `./foo` are *not*
+  # treated as equal. To compare paths semantically, they need to be normalized
+  # and converted to the same flavour.
+  #
+  # Comparison is case-sensitive for POSIX paths and case-insensitive for
+  # Windows paths.
+  def <=>(other : Path)
+    ord = @name.compare(other.@name, case_insensitive: windows?)
+    return ord if ord != 0
+
+    @type <=> other.@type
+  end
+
+  # Returns a path representing the drive component or `nil` if this path does not contain a drive.
+  #
+  # See `#anchor` for the combination of drive and `#root`.
+  #
+  # ```
+  # Path.windows("C:\Program Files").drive     # => Path.windows("C:")
+  # Path.windows("\\host\share\folder").drive  # => Path.windows("\\host\share")
+  # ```
+  #
+  # NOTE: Drives are only available for Windows paths. It can either be a drive letter (`C:`) or a UNC share (`\\host\share`).
+  def drive : Path?
+    if drive = @type.drive(@name)
+      new_instance drive
+    end
+  end
+
+  # Returns the root path component of this path or `nil` if it is not rooted.
```

Comment:
What's the usecase for this?

Reply:
I'm not entirely sure myself. But both Java and Python API's have this, so I'd guess it's probably useful for something... internally it's used by `#anchor`.